webauthn-security-key/server.js

```diff
@@ -0,0 +1,116 @@
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+
+const express = require('express');
+const session = require('express-session');
+const hbs = require('hbs');
+const auth = require('./libs/auth');
+const app = express();
+
+app.set('view engine', 'html');
+app.engine('html', hbs.__express);
+app.set('views', './views');
+app.use(express.json());
+app.use(express.static(__dirname + '/public'));
+app.use(express.static('dist'));
+
+app.use(
+  session({
+    // You should specify a real secret here
+    secret: 'secret',
+    resave: true,
+    saveUninitialized: false,
+    proxy: true,
+    cookie: {
+      httpOnly: true,
+      secure: true,
+      sameSite: 'none',
+    },
+  })
+);
+
+function isFullySignedIn(req) {
+  return req.session.name === 'main';
+}
+
+function isPartlySignedIn(req) {
+  return req.session.username && req.session.isPasswordCorrect;
+}
+
+app.use((req, res, next) => {
+  if (process.env.PROJECT_DOMAIN) {
+    process.env.HOSTNAME = `${process.env.PROJECT_DOMAIN}.glitch.me`;
+  } else {
+    process.env.HOSTNAME = req.headers.host;
+  }
+  const protocol = /^localhost/.test(process.env.HOSTNAME) ? 'http' : 'https';
+  process.env.ORIGIN = `${protocol}://${process.env.HOSTNAME}`;
+  if (
+    req.get('x-forwarded-proto') &&
+    req.get('x-forwarded-proto').split(',')[0] !== 'https'
+  ) {
+    return res.redirect(301, process.env.ORIGIN);
+  }
+  req.schema = 'https';
+  next();
+});
+
+app.get('/', async (req, res) => {
+  if (isFullySignedIn(req)) {
+    // If the user is signed in, redirect to the account page
+    res.redirect(307, '/account');
+    return;
+  }
+  // If the user is not signed in, start a new "auth" session
+  try {
+    // Start the intermediate session "auth", dedicated to authentication/login
+    req.session.name = 'auth';
+    // "auth" expires after 3 minutes, this means the user has 3 minutes to authenticate
+    const sessionLength = 3 * 60 * 1000;
+    req.session.cookie.expires = new Date(Date.now() + sessionLength);
+    // Render the index page
+    res.render('index.html');
+  } catch (e) {
+    console.log(e);
+  }
+});
+
+app.get('/account', (req, res) => {
+  if (!isFullySignedIn(req)) {
+    // If user is not fully signed in, redirect to the index page with the login form
+    res.redirect(307, '/');
+    return;
+  }
+  res.render('account.html', { username: req.session.username });
+});
+
+app.get('/secondFactor', (req, res) => {
+  if (!isPartlySignedIn(req)) {
+    // If user is not signed in, redirect to the index page with the login form.
+    res.redirect(302, '/');
+    return;
+  }
+  if (isFullySignedIn(req)) {
```

Comment:
@agektmr Terminology questions: 
1. In the code: I'm using "signed-in" and "authenticated" interchangeably in function names. I'd like to unify this at least in the code, WDYT?
2. In the UI: same, I'm using both terms. Should we stick to one of these terms? Or is it OK, since "authenticate" is traditionally used for "(2F)A" and "sign in" is a well-known term?

Reply:
Use one in the code (e.g. signed in)

Reply:
@agektmr I went for "auth" since it had many more occurences. All function names now use "auth", only the UI uses "sign in" since it's a more common term for users.

Reply:
Makes sense.